WalletWasabi/BitcoinCore/P2pNode.cs

```diff
@@ -0,0 +1,214 @@
+using NBitcoin;
+using NBitcoin.Protocol;
+using Nito.AsyncEx;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.BitcoinCore.Monitoring;
+using WalletWasabi.Blockchain.Blocks;
+using WalletWasabi.Blockchain.Mempool;
+using WalletWasabi.Blockchain.P2p;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BitcoinCore
+{
+	public class P2pNode
+	{
+		private Node Node { get; set; }
+		private TrustedP2pBehavior TrustedP2pBehavior { get; set; }
+		public Network Network { get; }
+		public EndPoint EndPoint { get; }
+		public MempoolService MempoolService { get; }
+
+		public event EventHandler<uint256> BlockInv;
+
+		private bool NodeEventsSubscribed { get; set; }
+		private object SubscriptionLock { get; }
+
+		private CancellationTokenSource Stop { get; set; }
+		private P2pReconnector Reconnector { get; set; }
+		private AsyncLock ReconnectorLock { get; }
+
+		public P2pNode(Network network, EndPoint endPoint, MempoolService mempoolService, string userAgent)
+		{
+			Network = Guard.NotNull(nameof(network), network);
+			EndPoint = Guard.NotNull(nameof(endPoint), endPoint);
+			MempoolService = Guard.NotNull(nameof(mempoolService), mempoolService);
+			UserAgent = Guard.NotNullOrEmptyOrWhitespace(nameof(userAgent), userAgent, trim: true);
+
+			Stop = new CancellationTokenSource();
+			NodeEventsSubscribed = false;
+			SubscriptionLock = new object();
+			Reconnector = null;
+			ReconnectorLock = new AsyncLock();
+		}
+
+		public async Task ConnectAsync(CancellationToken cancel)
+		{
+			using var handshakeTimeout = new CancellationTokenSource();
+			using var linked = CancellationTokenSource.CreateLinkedTokenSource(handshakeTimeout.Token, cancel, Stop.Token);
+			handshakeTimeout.CancelAfter(TimeSpan.FromSeconds(21));
+			var parameters = new NodeConnectionParameters()
+			{
+				UserAgent = UserAgent,
+				ConnectCancellation = linked.Token,
+				IsRelay = true
+			};
+
+			parameters.TemplateBehaviors.Add(new TrustedP2pBehavior(MempoolService));
+
+			Node = await Node.ConnectAsync(Network, EndPoint, parameters).ConfigureAwait(false);
+			Node.VersionHandshake();
```

Comment:
Can be, but I don't know how. Feel free to show me. I kinda feel it pointless, because they should not use different Core version from what Wasabi uses in the first place.